adr/0011-python-application-deployment.md

```diff
@@ -0,0 +1,77 @@
+checks. These tasks are often glued together by the CI runner, or bash scripts or other python code and, recently, Make 
+files.
+
+In the case of projects with multiple authors, there are often multiple ways present to achieve the same result that 
+suits individual authors best.
+
+Python projects have traditionally left deployment and configuration up to the project formula. The project formula
+describes the requirements of the environment the project lives in, for example, does a database or web server need to 
+exist? What directories do log files live in and what permissions do they need? These and dozens of other similar cases
+are handled by the formula on behalf of the Python project. The project author doesn't necessarily need to know about 
+how these cases are solved, the formula is built and maintained by those building and maintaining the systems these
+projects run on in production.
+
+Python projects require a Python runtime and Python packages can be installed system-wide or locally using a tool called
+'virtualenv', now officially part of Python3. Dependency and runtime management has become complicated in the move from
+Python2 to Python3 and the introduction of new language features and dependency management tools; the OS distributions 
+used in production use a fixed version of the Python runtime, often with features missing or mutually exclusive with 
+other features (Ubuntu + Python3 + pip + venv for example). Attempts to accommodate the latest runtimes has introduced
+instability and complexity in the rest of the system.
+
+Containerisation has made describing and building Python applications much easier for a project's author and the line 
+between traditional code development and code deployment is less clear. Container descriptions often live within the 
+code repository instead of in a separate one. Containers serve a single purpose, running a single process, but are 
+composable so that a cluster of containers may be described: one for the application, another for a webserver, another
+for a database and so on.
+
+Deploying a container that has the project within perfectly installed and configured is very similar to deploying a fat
+application binary: a single discrete artifact with all dependencies accounted for. Deployment consists of dropping the
+artifact into the runtime and reloading a service.
+
+Because container descriptions live alongside the codebase and are used during development they suffer the same 
+problems as the author-preferred utility scripts/glue code. There are any number of ways to build a container, does 
+installation/testing/linting/etc happen within the container or without? Are the build commands invoked in the same way 
+across projects? If we insist on containers, then is it the project author's responsibility to now maintain a suite of 
+them? Are they the best party to attempt it.
+
+## Decision
+
+Deploy all Python projects within containers.
```

Comment:
This seems uncontroversial, at least for new projects, and not just related to Python but also PHP or Node.js.

Reply:
Perhaps trivial but the container definition (e.g. `Dockerfile`) should be inside the project repository? As opposed to in anything else like a formula repository

Reply:
Could have sworn we had an ADR for containerising-all-the-things given that's what we've been doing, but apparently not. 🤔

Reply:
> Could have sworn we had an ADR for containerising-all-the-things

I haven't seen that ADR come across my radar. I've been bugging Giorgio (who seems to have more of an idea than I do when it comes to deploying these containers) just what our policy and roadmap should be.

> Perhaps trivial but the container definition (e.g. Dockerfile) should be inside the project repository? 

I don't want to tell devs how to develop, but they seem to like their Dockerfiles living alongside their code in the project repository, so I say keep it there. I mention this in the ADR but perhaps it wasn't clear (or too word-y). Unfortunately, that Dockerfile will be used for build+test+deploy, so if they want a different one for themselves/dev they can create a prefixed `-dev` Dockerfile or submit to the decisions in the standard one.

Reply:
I agree containers are good, not that many of the repos I'm working on are structured that way right now.

When you use the term `"projects"` here, does it also include things we sometimes refer to as `"libraries"`, where the only executable part of them are the test cases themselves?

Reply:
I think if you read "projects" as "project repositories" that may help. Yes, it would apply to libraries as well like, like `elife-tools` as they have extensive tests and non-python dependencies, however code that depends on those libraries, like bot-lax-adaptor or elife-bot, wouldn't use the containerised form.